makeOverridablePythonPackage: simplify implementation and make compatible with `<pkg>.overrideAttrs`

ShamrockLee opened the pull request:
## Description of changes

This patch updates the `<pkg>.overrideAttrs` inside `makeOverridablePythonPackage` (the function that adds `<pkg>.overridePythonAttrs`). If applied, `<pkgs>.overridePythonAttrs` will work after applying `<pkg>.overrideAttrs` to override a Python package built with `buildPythonPackage` or `buildPythonApplication`.

This patch also adds corresponding test cases to `tests.overriding`.

Closes #267293.

Initial effort for #271387.

Cc: @FRidh 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
  - [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ShamrockLee committed:
tests.overriding: modularize Python tests

Test with a pseudopackage python-package-stub.

Modularize the transforming function and testing function.

ShamrockLee committed:
makeOverridablePythonPackage: take care of overrideAttrs

Make it possible to mix overridePythonAttrs and overrideAttrs, i.e.
((<pkg>.overrideAttrs (_: { foo = "a"; })).overridePythonAttrs (_: { })).foo now works

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

ShamrockLee commented:
@ofborg build tests.overriding

ShamrockLee commented:
@ofborg build tests.overriding

ShamrockLee commented:
@ofborg build tests.overriding

ShamrockLee commented:
@ofborg build python3Packages.xpybutil

ShamrockLee commented:
@ofborg build tests.overriding

SuperSandro2000 commented on pkgs/test/overriding.nix:
```diff
@@ -3,12 +3,42 @@
   tests =
     let
-      p = pkgs.python3Packages.xpybutil.overridePythonAttrs (_: { dontWrapPythonPrograms = true; });
+      inherit (pkgsAllowingBroken.python3Packages) pillow xpybutil;
```
Is the choice important here? If so I would use something like pip or maybe cryptography but not an orchide like pillow or xpyutil.

ShamrockLee replied:
> Is the choice important here? If so I would use something like pip or maybe cryptography but not an orchide like pillow or xpyutil.

`xpyutil` appears in a previous test case, and `pillow` is one of its dependencies. There's no special reason to choose them. `pip` and `cryptography` sounds like a good idea, as they are not expected to break.

ShamrockLee replied:
Addressed.

SuperSandro2000 commented on pkgs/test/overriding.nix:
```diff
@@ -3,12 +3,42 @@
     let
-      p = pkgs.python3Packages.xpybutil.overridePythonAttrs (_: { dontWrapPythonPrograms = true; });
+      inherit (pkgsAllowingBroken.python3Packages) pillow xpybutil;
+      applyOverridePythonAttrs = p: p.overridePythonAttrs (_: {dontWrapPythonPrograms = true; });
+      revertOverridePythonAttrs = p: p.overridePythonAttrs (_: {dontWrapPythonPrograms = false; });
```
```suggestion
      applyOverridePythonAttrs = p: p.overridePythonAttrs (_: { dontWrapPythonPrograms = true; });
      revertOverridePythonAttrs = p: p.overridePythonAttrs (_: { dontWrapPythonPrograms = false; });
```

SuperSandro2000 commented on pkgs/development/interpreters/python/python-packages-base.nix:
```diff
@@ -30,13 +33,17 @@ let
-        __functor = self: result;
-      }
-      else result);
+    if builtins.isAttrs result then result
+      // lib.optionalAttrs (result ? overrideAttrs) { overrideAttrs = fdrv: overrideResult (drv: drv.overrideAttrs fdrv); }
```
Don't we need () here or are the redundant?

ShamrockLee replied:
I don't quite get it.

`result ? overrideAttrs` would be true only when `result` is an attribute. Maybe

```nix
if builtins.isAttrs result then
  result // lib.optionalAttrs (result?overrideAttrs) {
    overrideAttrs =...;
  }
else if ...
```

If you are talking about the let-in construct above the if-then-else construct, it surely doesn't need parenthesis.

```
nix-repl> let in if 2 + 2 == 5 then "y" else "n"
"n"
```

ShamrockLee replied:
> Don't we need () here or are the redundant?

Or do you mean that `set?attr` has higher precedence over function application and the parenthesis around it is redundant?

ShamrockLee commented:
I rebased the feature branch and resolved the merge conflict.

MattSturgeon commented on pkgs/development/interpreters/python/python-packages-base.nix:
```diff
@@ -14,26 +14,40 @@ let
         result = f args;
-        overrideWith = newArgs: args // (if pkgs.lib.isFunction newArgs then newArgs args else newArgs);
+
+        overrideWith = newArgs: args // lib.toFunction newArgs args;
```
Slightly off topic:

`overrideAttrs` supports being called with:
1. `attrs`
2. `(prev: attrs)`
3. `(final: prev: attrs)`

At some point, it'd be nice if `overridePythonAttrs` also supported the third style (overlay-style extension function), currently it only supports the first two.

I assume it'd be possible to encapsulate that within `overrideWith`? In another PR, of course.

ShamrockLee replied:
It can be done in `overrideWith`, but I don't feel like letting `overridePythonAttrs` take fixed-point arguments.
1. `overridePythonAttrs` is a custom overrider and `buildPythonPackage` argument mapping contains a lot of historical burden (e.g., `checkPhase` -> `installCheckPhase`). I'm leaning toward deprecating `buildPythonPackage` after fixing the `finalAttrs` reference for `<pkg>.overrideAttrs` and fixed-point arguments.
2. We need to subtract the `stdenv` argument as implemented in #271762, and supporting fixed-point arguments make it much more challenging (if not impossible to implement).

ShamrockLee commented on pkgs/development/interpreters/python/python-packages-base.nix:
```diff
@@ -14,26 +14,40 @@ let
         result
+        // lib.optionalAttrs (result ? overrideAttrs) {
+          overrideAttrs = fdrv: overrideResult (drv: drv.overrideAttrs fdrv);
+        }
```
That will surely make `makeOverridablePythonAtttrs` implementation cleaner more understandable.

Still, I'm hesitant to apply risky improvements to something we will eventually deprecate.

ShamrockLee committed:
makeOverridablePythonPackage: simplify implementation

Attach overridePythonAttrs by attribute set update
instead of `stdenv.mkDerivation`'s `passthru`.

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

MattSturgeon commented on pkgs/test/overriding.nix:
```diff
@@ -7,6 +7,15 @@
+        allowBroken = true;
+      };
+    }
+  );
```
Need to update the commit message:

> Test with pip, which is less likely to fail.

MattSturgeon commented on pkgs/development/interpreters/python/python-packages-base.nix:
```diff
@@ -14,27 +14,24 @@ let
+        result = f origArgs;
+        overrideWith = newArgs: origArgs // lib.toFunction newArgs origArgs;
+        # Change the result of the function call by applying `transform` to it
+        overrideResult = transform: makeOverridablePythonPackage (args: transform (f args)) origArgs;
```
Nit: "override result" is a noun that doesn't really convey that this is a function; it sounds more like a static value. I guess it could mean "override the result", but I interpreted it as "the result of overriding".

We could name it something like "wrapOverride" or "applyTransform"?

ShamrockLee commented:
Rebase and utilize PR #456848 ("tests.overriding: refactor and allow non-bool expected value").

ShamrockLee commented on pkgs/development/interpreters/python/python-packages-base.nix:
```diff
@@ -14,27 +14,24 @@ let
+        // {
           overridePythonAttrs = newArgs: makeOverridablePythonPackage f (overrideWith newArgs);
-          __functor = self: result;
+          overrideAttrs = newArgs': overrideResult (drv: drv.overrideAttrs newArgs');
```
Addressed.

ShamrockLee commented:
I wonder if it's okay to backport this PR, especially after the `makeOverridablePythonAttrs` refactoring (https://github.com/NixOS/nixpkgs/pull/267296#discussion_r2452289710).

It should fix `<pkg>.overridePythonAttrs` and enable split tests implemented with `<pkg>.overrideAttrs` due to `buildPythonPackage`'s lack of fixed-point attribute support. PR #469544 is an example of such pattern.

MattSturgeon reviewed (approved):
> I wonder if it's okay to backport this PR

I believe this should be fine to backport. It doesn't do anything too crazy.

---

Diff LGTM.
